src/main/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporter.java

```diff
@@ -0,0 +1,341 @@
+        } else if (event instanceof AuctionEvent auctionEvent) {
+            greenbidsAuctionContext = auctionEvent.getAuctionContext();
+            greenbidsBidResponse = auctionEvent.getBidResponse();
+        } else {
+            greenbidsAuctionContext = null;
+            greenbidsBidResponse = null;
+        }
+
+        if (greenbidsBidResponse == null || greenbidsAuctionContext == null) {
+            Future.failedFuture("Bid response or auction context cannot be null");
+        }
+
+        final String greenbidsId = UUID.randomUUID().toString();
+        final String billingId = UUID.randomUUID().toString();
+
+        isSampled(greenbidsAnalyticsProperties.getGreenbidsSampling(), greenbidsId)
+                .compose(isSampled -> {
+                    if (isSampled) {
+                        final Future<CommonMessage> bidMessage = createBidMessage(
+                                greenbidsAuctionContext, greenbidsBidResponse, greenbidsId, billingId);
+
+                        return bidMessage.onSuccess(commonMessage -> {
+                            try {
+                                final String commonMessageJson = jacksonMapper.encodeToString(commonMessage);
+                                HttpUtil.sendJson(commonMessageJson, greenbidsAnalyticsProperties.getAnalyticsServer());
+                            } catch (EncodeException e) {
+                                Future.failedFuture("Failed to encode as JSON: " + e.getMessage());
+                            }
+                        });
+                    }
+                    return null;
+                });
+
+        return Future.succeededFuture();
+    }
+
+    private Future<Boolean> isSampled(
+            double samplingRate,
+            String greenbidsId) {
+        final Promise<Boolean> promise = Promise.promise();
+        try {
+            if (samplingRate < 0 || samplingRate > 1) {
+                Future.failedFuture("Warning: Sampling rate must be between 0 and 1");
+            }
+
+            final double exploratorySamplingRate = samplingRate
+                    * greenbidsAnalyticsProperties.getExploratorySamplingSplit();
+            final double throttledSamplingRate = samplingRate
+                    * (1.0 - greenbidsAnalyticsProperties.getExploratorySamplingSplit());
+
+            final long hashInt = Math.abs(greenbidsId.hashCode()) % RANGE_16_BIT_INTEGER_DIVISION_BASIS;
+            final boolean isPrimarySampled = hashInt < exploratorySamplingRate * MAX_16_BIT_INTEGER;
+
+            promise.complete(isPrimarySampled || hashInt >= (1 - throttledSamplingRate) * MAX_16_BIT_INTEGER);
+
+        } catch (IllegalArgumentException e) {
+            promise.fail(e);
+        }
+
+        return promise.future();
+    }
+
+    public Future<CommonMessage> createBidMessage(
+            AuctionContext auctionContext,
+            BidResponse bidResponse,
+            String greenbidsId,
+            String billingId) {
```

Comment:
same, it's no necessary that each method returns future

also why is the method public?

Reply:
Have reverted changes for `createBidMessage` returning futures.

I made this method public because I then call it in unit test `GreenbidsAnalyticsReporterTest`. This is because in unit test I check the content of the payload without sending it to analytics server.

Reply:
Otherwise if I make the method `createBidMessage` private then in UT I will need to call `processEvent` instead and mock both creation of a payload and send to the mocked server.

The analytics server could be mocked using [PowerMock](https://github.com/powermock/powermock) but in PBS I did not find that it was used before